### PR TITLE
Alternate communications mechanism for hosts that block xmlrpc.php

### PR DESCRIPTION
**Pull request by gravityrail**
Fixes an issue where sites could not connect from hosts that blanket-block access to XMLRPC.

#### Changes proposed in this Pull Request:

* expose ONLY Jetpack methods (not default XMLRPC API) via new route, as a fallback

#### Testing instructions:

* Later

### CODE REVIEW

File: _inc/lib/class.core-rest-api-endpoints.php [hunk 1]
```diff
@@ -304,6 +310,21 @@ public static function register_endpoints() {
...
+	 * @since 4.9.0
+	 *
+	 * @param WP_REST_Request $request The request sent to the WP REST API.
+	 *
+	 * @return array|wp-error
+	 */
+	public static function verify_registration( $request ) {
+        require_once JETPACK__PLUGIN_DIR . 'class.jetpack-xmlrpc-server.php';
+		$xmlrpc_server = new Jetpack_XMLRPC_Server();
+		return $xmlrpc_server->verify_registration( array( $request['secret_1'], $request['state'] ) );
```
Comment (beaulebens) [inline _inc/lib/class.core-rest-api-endpoints.php]: Some wonky indentation going on here.

File: class.jetpack-data.php [hunk 2]
```diff
@@ -93,22 +93,6 @@ public static function is_usable_domain( $domain, $extra = array() ) {
...
-		// Check the IP to make sure it's pingable.
-		$ip = gethostbyname( $domain );
-
-		// Doing this again as I was getting some false positives when gethostbyname() flaked out and returned the domain.
-		$ip = filter_var( $ip, FILTER_VALIDATE_IP, FILTER_FLAG_IPV4 ) ? $ip : gethostbyname( $ip );
-
-		if ( ! filter_var( $ip, FILTER_VALIDATE_IP, FILTER_FLAG_NO_PRIV_RANGE | FILTER_FLAG_NO_RES_RANGE | FILTER_FLAG_IPV4 ) && ! self::php_bug_66229_check( $ip ) ) {
-			return new WP_Error( 'fail_domain_bad_ip_range', sprintf( __( 'Domain `%1$s` just failed is_usable_domain check as its IP `%2$s` is either invalid, or in a reserved or private range.', 'jetpack' ), $domain, $ip ) );
-		}
-
```
Comment (beaulebens) [inline class.jetpack-data.php]: Why is all of this section removed? And alongside that... why did we ever need it in the first place? :)

File: class.jetpack-xmlrpc-fallback.php [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        register_rest_route( 'jetpack/v1', '/xmlrpc', array(
+			'methods' => 'POST',
+			'callback' => array( $this, 'xmlrpc_fallback' ),
+		) );
+    }
+
+    function verify_registration( $request ) {
+        $this->load_xmlrpc();
+		return $this->xmlrpc_server->verify_registration( array( $request['secret_1'], $request['state'] ) );
+	}
```
Comment (beaulebens) [inline class.jetpack-xmlrpc-fallback.php]: Wonky indentation again.

File: class.jetpack-xmlrpc-fallback.php [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+    // allow invocation of Jetpack XMLRPC, and a handful of core XMLRPC methods, over REST
+    // does NOT allow random XMLRPC requests
+    function xmlrpc_fallback( $request ) {
+        $this->load_xmlrpc();
+        $methods = $this->xmlrpc->xmlrpc_methods();
+        error_log("xmlrpc fallback");
+        error_log(print_r($methods,1));
```
Comment (beaulebens) [inline class.jetpack-xmlrpc-fallback.php]: This is incomplete, right? :)

File: class.jetpack.php [hunk 5]
```diff
@@ -599,6 +611,55 @@ private function __construct() {
...
+
+		/**
+		 * Filters the class used for handling XML-RPC requests.
+		 *
+		 * @since 3.1.0
+		 *
+		 * @param string $class The name of the XML-RPC server class.
+		 */
+		$wp_xmlrpc_server_class = apply_filters( 'wp_xmlrpc_server_class', 'wp_xmlrpc_server' );
+		$wp_xmlrpc_server = new $wp_xmlrpc_server_class;
```
Comment (beaulebens) [inline class.jetpack.php]: If we leave this in here, then I think we'll need to do some forcing to make sure that other methods can't be filtered into this XML RPC server.

**Comment by roccotripaldi** [comment]
Pinged you on #7115 and D5453 which should clear up some of the connection errors we are seeing. I'm not saying we shouldn't go this route as a fallback for connections, but #7115 feels like a eureka moment for secret mismatch errors.

File: class.jetpack.php [hunk 6]
```diff
@@ -479,7 +479,19 @@ private function __construct() {
 		add_action( 'deleted_user', array( $this, 'unlink_user' ), 10, 1 );
 		add_action( 'remove_user_from_blog', array( $this, 'unlink_user' ), 10, 1 );
 
+		// Alternate XML-RPC, via ?jetpack=comms
+		if ( isset( $_GET['jetpack'] ) && 'comms' == $_GET['jetpack'] ) {
+			if ( ! defined( 'XMLRPC_REQUEST' ) ) {
+				define( 'XMLRPC_REQUEST', true );
+			}
+
+			error_log("alternate xmlrpc");
```
Comment (lezama) [inline class.jetpack.php]: error_log

**Comment by gravityrail** [comment]
This has been closed in favor of #7740

**Pull request closed without merging**